packages/react-data/src/fetch/ReactDataFetch.js

```diff
@@ -44,50 +46,99 @@ function readResultMap(): Map<string, Result> {
-export function fetch(url: string): Object {
-  const map = readResultMap();
-  const entry = map.get(url);
-  if (entry === undefined) {
-    let resolve = () => {};
-    const wakeable: Wakeable = new Promise(r => {
-      // TODO: should this be a plain thenable instead?
-      resolve = r;
-    });
-    const result: Result = {
-      status: Pending,
-      value: wakeable,
-    };
-    map.set(url, result);
-    const xhr = new XMLHttpRequest();
-    xhr.onload = function() {
-      // TODO: should we handle status codes?
-      if (result.status !== Pending) {
-        return;
+function createFromThenable(thenable, wrapValue): Result {
+  const result: Result = {
+    status: Pending,
+    value: (null: any),
+  };
+  result.value = thenable.then(
+    value => {
+      if (result.status === Pending) {
+        const resolvedResult = ((result: any): ResolvedResult);
+        resolvedResult.status = Resolved;
+        resolvedResult.value = wrapValue(value);
       }
-      const resolvedResult = ((result: any): ResolvedResult);
-      resolvedResult.status = Resolved;
-      resolvedResult.value = xhr.response;
-      resolve();
-    };
-    xhr.onerror = function() {
-      if (result.status !== Pending) {
-        return;
+    },
+    err => {
+      if (result.status === Pending) {
+        const rejectedResult = ((result: any): RejectedResult);
+        rejectedResult.status = Rejected;
+        rejectedResult.value = err;
       }
-      const rejectedResult = ((result: any): RejectedResult);
-      rejectedResult.status = Rejected;
-      // TODO: use something else as the error value?
-      rejectedResult.value = xhr;
-      resolve();
-    };
-    xhr.open('GET', url);
-    xhr.send();
-    throw wakeable;
-  }
-  const result: Result = entry;
+    },
+  );
+  return result;
+}
+
+function readResult(result: Result) {
   if (result.status === Resolved) {
     return result.value;
   } else {
     throw result.value;
   }
 }
+
+function getJson(nativeResponse) {
+  return nativeResponse.json();
+}
+
+function getText(nativeResponse) {
+  return nativeResponse.json();
+}
+
+function Response(nativeResponse) {
+  this.headers = nativeResponse.headers;
+  this.ok = nativeResponse.ok;
+  this.redirected = nativeResponse.redirected;
+  this.status = nativeResponse.status;
+  this.statusText = nativeResponse.statusText;
+  this.type = nativeResponse.type;
+  this.url = nativeResponse.url;
+
+  this._entry = null;
+  this._consumed = null;
+  this._response = nativeResponse;
+}
+
+Response.prototype = {
+  json() {
+    return this._read('json', getJson);
+  },
+  text() {
+    return this._read('text', getText);
+  },
+  _read(type: string, getThenable) {
+    const consumed = this._consumed;
+    if (consumed != null && type !== consumed) {
+      throw new Error('Already read.');
+    }
+    let entry = this._entry;
+    if (entry == null) {
+      this._consumed = type;
+      const thenable = getThenable(this._response);
```

Comment:
Passing first class functions break inlining and what not. Just make a switch in here.

I'd just duplicate the code instead though and then extract from the inside out. E.g. make make this a helper function and call that in both functions: 

```
    const consumed = this._consumed;
    if (consumed != null && type !== consumed) {
      throw new Error('Already read.');
    }
    this._consumed = type;
```

Reply:
I was code golfing here a bit because the byte size of this helper seems important. Maybe not so much.